src/snowflake/cli/_plugins/nativeapp/entities/application.py

```diff
@@ -87,6 +105,177 @@
+        ) or []
+
+        self._metrics.set_counter(
+            CLICounterField.EVENT_SHARING, int(self._authorize_event_sharing or False)
+        )
+        self._metrics.set_counter(CLICounterField.EVENT_SHARING_WARNING, 0)
+        self._metrics.set_counter(CLICounterField.EVENT_SHARING_ERROR, 0)
+
+        if not self._event_sharing_enabled:
+            # We cannot set AUTHORIZE_TELEMETRY_EVENT_SHARING to True or False if event sharing is not enabled,
+            # so we will ignore the field in both cases, but warn only if it is set to True
+            if self._authorize_event_sharing or self._optional_shared_events:
+                console.warning(
+                    "WARNING: Same-account event sharing is not enabled in your account, therefore, application telemetry section will be ignored."
+                )
+            self._authorize_event_sharing = None
+            self._optional_shared_events = []
+            return
+
+        self._manifest_events_definitions = []
+        if not install_method.is_dev_mode:
+            # We can't make any decision if we are in prod mode because we do not know the manifest.
+            return
+
+        all_events_in_manifest = find_events_in_manifest_file(deploy_root_path)
+        mandatory_events_in_manifest = find_events_in_manifest_file(
+            deploy_root_path, mandatory_only=True
+        )
+        self._manifest_events_definitions = self._to_events_definitions(
+            all_events_in_manifest, mandatory_events_in_manifest
+        )
+
+        if self._optional_shared_events:
+            for event in self._optional_shared_events:
+                if event not in all_events_in_manifest:
+                    self._metrics.set_counter(CLICounterField.EVENT_SHARING_ERROR, 1)
+                    raise ClickException(
+                        f"Shared event '{event}' is not found in the manifest file."
+                    )
+
+        if mandatory_events_in_manifest and self._event_sharing_enforced:
+            if self._authorize_event_sharing is None:
+                self._metrics.set_counter(CLICounterField.EVENT_SHARING_WARNING, 1)
+                console.warning(
+                    "WARNING: Mandatory events are present in the manifest file. Automatically authorizing event sharing in dev mode. To suppress this warning, please add 'share_mandatory_events: true' in the application telemetry section."
+                )
+                self._authorize_event_sharing = True
+
+    def should_authorize_event_sharing(
+        self,
+        app_properties: Optional[Dict[str, str]] = None,
+        events_definitions: Optional[List[Dict[str, str]]] = None,
+    ) -> Optional[bool]:
+        if not self._event_sharing_enabled:
+            return None
+
+        if events_definitions is None and self._is_dev_mode:
+            events_definitions = self._manifest_events_definitions
+
+        current_authorization = (
+            app_properties
```

Comment:
for upgrade, the flag change happens after the upgrade. I was checking that we are not changing the flag from true to false when we actually have mandatory events already authorized. This will fail anyways.

In new rev, I simplified the logic so we rely on error codes from the backend instead of doing all the checks ourselves.